New documentation content (#671)
graknlabs/docs: 0001-typedb/02-dev/03-match.md

---
pageTitle: Match query
keywords: typeql, query, match, pattern, statement, variable
longTailKeywords: typeql match, match get, query patterns, match clause, typeql variables
Summary: Targeting instances of data or schema types that match expressive patterns in TypeDB.
---

# Matching patterns

TypeDB uses patterns to read and write data in a database, and to read the types in its schema. 
In [TypeQL](../../11-query/00-overview.md), a `match`, `get`, `insert` or `delete` keyword followed by one or more 
patterns is a clause, with one or more clauses forming a query.

These clauses can be combined to create four different types of queries:

- Reading schema and data

  - [Get](05-read.md#get-query) (match-get) query — to read data or schema types.

- Writing data

  - [Insert](04-write.md#insert-query) (match-insert) query — to insert data.
  - [Delete](04-write.md#delete-query) (match-delete) query — to delete data.
  - [Update](04-write.md#update) (match-delete-insert) query — to delete and then insert data.

A `match` clause is optional for Insert queries, but required for Get, Delete and Update queries.

<div class="note">
[Note]

> While the `match` clause can be used to read types in a schema, the define and undefine clauses must be used to write 

Reviewer: ```suggestion
While the `match` clause can be used to read types from a schema, the define and undefine clauses must be used to write 
```
Is it obvious what "define and undefine clauses" are?

them.
</div>

<div class="note">
[Important]
If a `match` clause is not combined with a `get`, `insert` or `delete` clause, TypeDB will assume it is a get query and 
add an implicit `get` clause to return all variables defined in the `match` clause.
</div>

## Patterns overview

At the core of each query sits a query pattern that describes a subset of data of our particular interest. Here we 
examine the structure of query patterns closer. In general, patterns can be thought of as different arrangements of 
statement collections. TypeQL statements constitute the smallest building blocks of queries. Let’s have a close look 
at the constructs of a basic match clause.

![Statement structure](../../images/query/statement-structure.png)

- A TypeQL variable is prefixed with a dollar sign `$`. Most statements start with a variable (`V`) providing a concept

>   reference. For more information on variables please see Variables section.

Reviewer: Hyperlink.

  concepts that satisfy our statement.

- We mark the end of the statement with a semicolon (`;`).

## Variables

Variables can:

- reference both data and schema concepts.
- reference all (zero, one or many) concepts that comply with requirements described in a query patterns.

<div class="note">
[Important]
TypeQL is **declarative**. When writing TypeQL patterns for queries, we simply describe a **set of requirements** for 
the information we would like to match (retrieve to perform [get](05-read.md#get-query)/
[insert](04-write.md#insert-query)/[delete](04-write.md#delete-query)), rather than writing an algorithm of how 
should it be obtained. Once we specify the target information to match, the TypeQL query processor will take care of 
finding an optimal way to retrieve it.
</div>

There is some freedom in forming and composing our statements. For example, as shown below, we could write our single
statement with three properties:

```typeql
$p isa person, has full-name 'Masako Holley', has email $email;
```

However, you can also use separate patterns to achieve the same result:

```typeql
$p isa person;
$p has full-name 'Masako Holley';
$p has email $email;
```

> Lastly, to create a viable query from that statement, we need to add match keyword at the beginning. After that it is 

Reviewer: ```suggestion
Lastly, to create a viable query from that statement, we need to add the keyword `match` at the beginning. After that it is 
```

> ready to be sent to the TypeDB server in a data read transaction to a database with a corresponding schema.

Reviewer: ```suggestion
ready to be sent to the TypeDB server in a data read transaction to a database with an appropriate schema.
```

<div class="note">
[Note]

> If you want to run these and other examples from our documentation, make sure you have completed the Quickstart guide 

Reviewer: Add hyperlink.

to set up the database with relevant schema and dataset.
</div>

```typeql
match
  $p isa person;
  $p has full-name 'Masako Holley';
  $p has email $email;
```

<div class="note">
[Note]

> All indentation and line breaking is optional in TypeQL. You can even try using a single line or a complete random 

Reviewer: ```suggestion
All indentation and line breaking is optional in TypeQL. You can even try using a single line or a completely random 
```

| comparison                          | `$<var_label> [<comparison-operator>] <value>`                                                                                 |
| equal                               | `$<var_label> is $<var_label>`                                                                                                 |

### Comparison operators

The following operators are supported for comparing attribute values: `=`, `!=`, `>`, `>=`, `<` and `<=`.

### Combining patterns

![Combining statements](../../images/query/pattern-structure.png)

By arranging statements together, we can express more complex pattern scenarios and their corresponding data.

- Statement: simplest possible arrangement — a single basic building block as [explained above](#patterns-overview).
- Conjunction: a set of patterns where to satisfy a match, all patterns must be matched. We use conjunctions by default 
  just by separating the partaking patterns with semicolons `;`.
- Disjunction: a set of patterns where to satisfy a match, at least one pattern must be matched. We form

>   disjunctions by enclosing the partaking patterns within curly braces {} and interleaving them with the or keyword.

Reviewer: ```suggestion
  disjunctions by enclosing the relevant patterns within curly braces {} and joining them together with the keyword `or`.
```

> - Negation: defines a conjunctive pattern that explicitly defines conditions not to be met. We form negations by 

Reviewer: ```suggestion
- Negation: a conjunctive pattern that explicitly defines conditions that must not be met. We form negations by 
```

>   defining the pattern of interest inside a not {}; block.

Reviewer: ```suggestion
  defining the pattern of interest inside a `not {};` block.
```

## Match clause

### Schema queries

A `match` clause can be used to find types or roles in a database schema.

#### Type matching

##### All types

> Use a `subtype` pattern with type `thing` to find all types defined in a schema.

Reviewer: This will change in 3.0, so we might want to highlight that somehow.

Author: How do we produce the same result? 
I mean
```
match {$t sub entity;} or {t sub attribute;} or {t sub relation;};
```
doesn't work.

Author: NVM, I have found a workaround:
```
match $t sub $a;
```
